File: pt-2/851.loud-and-rich.java
import java.util.*;

/**
 * Algorithm:
 * Build edges from a poorer person to richer people. Memoized DFS returns the
 * quietest reachable richer-or-equal person for each node in the DAG.
 *
 * Java data structures:
 * ArrayList<Integer>[] is the adjacency list. int[] ans stores memoized
 * answers, using -1 as the uncomputed sentinel.
 *
 * Complexity:
 * Time O(n + e), space O(n + e).
 */
class Solution {
    private List<Integer>[] richerThan;
    private int[] quiet;
    private int[] ans;

    public int[] loudAndRich(int[][] richer, int[] quiet) {
        int n = quiet.length;
        this.quiet = quiet;
        richerThan = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            richerThan[i] = new ArrayList<>();
        }
        for (int[] edge : richer) {
            int rich = edge[0];
            int poor = edge[1];
            richerThan[poor].add(rich);
        }
        ans = new int[n];
        Arrays.fill(ans, -1);
        for (int person = 0; person < n; person++) {
            dfs(person);
        }
        return ans;
    }

    private int dfs(int person) {
        if (ans[person] != -1) {
            return ans[person];
        }
        int best = person;
        for (int rich : richerThan[person]) {
            int candidate = dfs(rich);
            if (quiet[candidate] < quiet[best]) {
                best = candidate;
            }
        }
        ans[person] = best;
        return best;
    }
}

File: Goldman Sachs/851.loud-and-rich.py
#
# @lc app=leetcode id=851 lang=python3
#
# [851] Loud and Rich
#
# https://leetcode.com/problems/loud-and-rich/description/
#
# algorithms
# Medium (63.45%)
# Likes:    1498
# Dislikes: 879
# Total Accepted:    78.2K
# Total Submissions: 123.3K
# Testcase Example:  '[[1,0],[2,1],[3,1],[3,7],[4,3],[5,3],[6,3]]\n[3,2,5,4,6,1,7,0]'
#
# There is a group of n people labeled from 0 to n - 1 where each person has a
# different amount of money and a different level of quietness.
# 
# You are given an array richer where richer[i] = [ai, bi] indicates that ai
# has more money than bi and an integer array quiet where quiet[i] is the
# quietness of the i^th person. All the given data in richer are logically
# correct (i.e., the data will not lead you to a situation where x is richer
# than y and y is richer than x at the same time).
# 
# Return an integer array answer where answer[x] = y if y is the least quiet
# person (that is, the person y with the smallest value of quiet[y]) among all
# people who definitely have equal to or more money than the person x.
# 
# 
# Example 1:
# 
# 
# Input: richer = [[1,0],[2,1],[3,1],[3,7],[4,3],[5,3],[6,3]], quiet =
# [3,2,5,4,6,1,7,0]
# Output: [5,5,2,5,4,5,6,7]
# Explanation: 
# answer[0] = 5.
# Person 5 has more money than 3, which has more money than 1, which has more
# money than 0.
# The only person who is quieter (has lower quiet[x]) is person 7, but it is
# not clear if they have more money than person 0.
# answer[7] = 7.
# Among all people that definitely have equal to or more money than person 7
# (which could be persons 3, 4, 5, 6, or 7), the person who is the quietest
# (has lower quiet[x]) is person 7.
# The other answers can be filled out with similar reasoning.
# 
# 
# Example 2:
# 
# 
# Input: richer = [], quiet = [0]
# Output: [0]
# 
# 
# 
# Constraints:
# 
# 
# n == quiet.length
# 1 <= n <= 500
# 0 <= quiet[i] < n
# All the values of quiet are unique.
# 0 <= richer.length <= n * (n - 1) / 2
# 0 <= ai, bi < n
# ai != bi
# All the pairs of richer are unique.
# The observations in richer are all logically consistent.
# 
# 
#

# @lc code=start
from collections import defaultdict
from typing import List


class Solution:
    def loudAndRich(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        richer_than = defaultdict(list)
        for rich, poor in richer:
            richer_than[poor].append(rich)

        ans = [-1] * len(quiet)

        def dfs(person: int) -> int:
            if ans[person] != -1:
                return ans[person]
            best = person
            for rich in richer_than[person]:
                candidate = dfs(rich)
                if quiet[candidate] < quiet[best]:
                    best = candidate
            ans[person] = best
            return best

        for person in range(len(quiet)):
            dfs(person)
        return ans
# @lc code=end

"""
Interview explanation:
For each person x, the answer is the quietest person among x and all ancestors in the richer-than DAG. Build edges from poorer person to richer people, then memoized DFS computes the best ancestor answer once per node.

Data structure: adjacency list represents the DAG; ans doubles as memoization.

Edge cases: people with no richer known person answer themselves. The input is acyclic by problem statement, so DFS does not need cycle handling.

Complexity: each node and richer relation is processed once, so O(n + e) time and O(n + e) space.
"""
